packages/loopring_v3/contracts/amm/libamm/AmmBlockReceiver.sol

```diff
@@ -21,73 +21,88 @@ library AmmBlockReceiver
     using AmmUpdateProcess  for AmmData.State;
     using BlockReader       for ExchangeData.Block;
 
-    function beforeBlockSubmission(
-        AmmData.State      storage S,
-        ExchangeData.Block memory  _block,
-        bytes              memory  poolTxData,
-        uint                       txIdx
-        )
+    function beforeAllBlocks(AmmData.State storage S)
         internal
-        returns (uint)
-    {
-        AmmData.Context memory ctx = _getContext(S, _block, txIdx);
-
-        BlockReader.BlockHeader memory header = _block.readHeader();
-        require(header.exchange == address(ctx.exchange), "INVALID_EXCHANGE");
-
-        S.approveAmmUpdates(ctx, true);
-
-        _processPoolTx(S, ctx, poolTxData);
-
-        S.approveAmmUpdates(ctx, false);
-
-        // Update state
-        S.poolTokenBurnedSupply = ctx.poolTokenBurnedSupply;
-
-        return ctx.txIdx - txIdx;
-    }
-
-    function _getContext(
-        AmmData.State      storage S,
-        ExchangeData.Block memory  _block,
-        uint                       txIdx
-        )
-        private
         view
         returns (AmmData.Context memory)
     {
         uint size = S.tokens.length;
         return AmmData.Context({
-            _block: _block,
-            txIdx: txIdx,
+            txIdx: 0,
             exchange: S.exchange,
             exchangeDomainSeparator: S.exchange.getDomainSeparator(),
             domainSeparator: S.domainSeparator,
             accountID: S.accountID,
             poolTokenID: S.poolTokenID,
-            poolTokenBurnedSupply: S.poolTokenBurnedSupply,
-            size: size,
             tokens: S.tokens,
-            tokenBalancesL2: new uint96[](size)
+            tokenBalancesL2: new uint96[](size),
+            poolTokenBurnedSupply: S.poolTokenBurnedSupply,
+            pendingTxIdx: 0,
+            pendingTxOwners: new address[](size * 3),
+            pendingTxHashes: new bytes32[](size * 3)
         });
     }
 
+    function afterAllBlocks(
+        AmmData.State   storage S,
+        AmmData.Context memory  ctx
+        )
+        internal
+    {
+        S.poolTokenBurnedSupply = ctx.poolTokenBurnedSupply;
+        S.exchange.approveTransactions(ctx.pendingTxOwners, ctx.pendingTxHashes);
+    }
+
+    function beforeEachBlock(
+        AmmData.State      storage /* S */,
+        ExchangeData.Block memory  _block,
+        AmmData.Context    memory  ctx
+        )
+        internal
+        pure
+    {
+        BlockReader.BlockHeader memory header = _block.readHeader();
+        require(header.exchange == address(ctx.exchange), "INVALID_EXCHANGE");
```

Comment:
Seems unlikely it's worth doing an extra call for these, especially if the functions are specifically written for AMM. Doing an external call (+ passing through the complete block) will be expensive. Will be cheaper/easier to just run over the contexts in `LoopringIOExchangeOwner` (which contains the exchange address the pool expects) and check them against the exchange addresses in the blocks there.